style: tentatively add hints for some public function

aarnphm opened the pull request:
This PR serves as a chore to tentatively introduce type hints into the library

It helps with development and catch bugs 😃

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

younesbelkada reviewed (commented):
Thanks a lot for adding this, this looks definitely great!
Can you resolve the merge conflicts? Also make sure to run the styling checks 
```bash
make style && make quality
```
After that we should be good to merge I think

aarnphm commented:
noted I will

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_614). All of your documentation changes will be reflected on that endpoint.

younesbelkada commented on src/peft/mapping.py:
```diff
@@ -13,6 +13,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from __future__ import annotations
```
Is that ever used below? Could you elaborate a bit on this import? 🙏

aarnphm replied:
Yes, in Python typing, all of the hints are evaluated at the import level. The below are how you usually do type hint:

```python
from Peft import PeftConfig

def something(config: PeftConfig): ...
```

The interpreter will know that the config of `something` will have type `PeftConfig`.
and its `__annotations__` for `something` would be:`{"config": PeftConfig, "return": None}`.

However, when imports are not used anywhere in the implementation other than hinting purposes, often times it is put under the `TYPE_CHECKING` block (as demonstrated in L39)

While static analysis tools such as pyright or mypy understand this, Python doesn't. Since Python will evaluate this type at runtime, this means it will lead to ImportError as the interpreter doesn't understand where `PeftConfig` is.

since [PEP563](https://peps.python.org/pep-0563/), the import `from __future__ import annotations` is added as a way for postponed evaluation for hinting. What this means is that now, the interpreter will evaluate all type hints within this given file as string

back to our `something` example, the hints when `__future__ import annotations` is available would become `{"config": "PeftConfig", "return": None}`. 
Tools such as pyright and mypy will know when it is evaluated as string, it will check the `TYPE_CHECKING` block for hints.

The benefits of evaluating types to string is that for modules or class that contains import side-effect, or just general takes a while to load, it plays nicely that we don't have to suffer importing it into the file while still have the full benefit of hinting.

Hope it helps.

aarnphm replied:
In addition to this, [PEP604](https://peps.python.org/pep-0604/) also allows us to write hint like this:
```python
Dict[str, Any] -> dict[str, Any]
```

PEP 604 is very objective, and since it is also relatively recent, I just use the conventional typing import to hint so that it is more readable for everyone who might not know about PEP 604 😃

younesbelkada replied:
Thanks for the great explanation ! Makes sense

pacman100 replied:
Thank you for the detailed information, helpful

younesbelkada reviewed (commented):
Thanks for your great contribution! I left one comment
Let's wait for @pacman100 before merging as usual

younesbelkada reviewed (approved):
Thanks a lot !

aarnphm commented:
I have fixed and addressed all change correspondingly. I believe #612 should get merged before this one.

So that I can fixed the merge conflict wrt to the imports
